boards/BMS/Src/App/states/App_PreChargeState.c

```diff
@@ -20,11 +49,67 @@ static void
     PreChargeStateRunOnTick100Hz(struct StateMachine *const state_machine)
 {
     App_AllStatesRunOnTick100Hz(state_machine);
+
+    struct BmsWorld *world = App_SharedStateMachine_GetWorld(state_machine);
+    struct Clock *   clock = App_BmsWorld_GetClock(world);
+    struct Airs *    airs  = App_BmsWorld_GetAirs(world);
+
+    float tractive_system_voltage = Io_VoltageSense_GetTractiveSystemVoltage();
+    float accumulator_voltage     = 355.0f; // TODO: Replace value with function to
+                                        // retrieve total pack voltage
+    uint32_t elapsed_time =
+        App_SharedClock_GetCurrentTimeInMilliseconds(clock) -
+        pre_charge_start_ms;
+
+    // if the voltage is not increasing there may be a SC across capacitor
+    if (tractive_system_voltage < accumulator_voltage &&
+        elapsed_time >= PRECHARGE_EXPECTED_CHARGE_TIME_MS * 1.20)
+    {
+        App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
+    }
+
+    // if the voltage increases too quickly there may be an OC before capacitor
+    else if (
+        tractive_system_voltage > accumulator_voltage &&
+        elapsed_time <= PRECHARGE_EXPECTED_CHARGE_TIME_MS * 0.20)
+    {
+        App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
+    }
```

Comment:
can probably do an OR here since they both transition into the fault state

Reply:
done